### PR TITLE
deprecate torch.testing.make_non_contiguous

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73348
* #72708
* __->__ #72705
* #72702

Differential Revision: [D34457731](https://our.internmc.facebook.com/intern/diff/D34457731)

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -1526,16 +1526,18 @@ def gradgradcheck(
...
-        # shape, type, and device as the outputs
-        def randn_like(x):
-            y = torch.randn_like(
-                x if (x.is_floating_point() or x.is_complex()) else x.double(), memory_format=torch.legacy_contiguous_format)
-            if gen_non_contig_grad_outputs:
-                y = torch.testing.make_non_contiguous(y)
-            return y.requires_grad_()
+        # If grad_outputs is not specified, create random Tensors of the same shape, type, and device as the outputs
+        def make_tensor_like(x):
+            return torch.testing.make_tensor(
```
Comment (albanD) [inline torch/autograd/gradcheck.py:1531]: How is this one different from the previous? 
In particular, things like layout or memory format are not copied anymore?

Comment (pmeier) [reply torch/autograd/gradcheck.py:1531]: > How is this one different from the previous?

By default `make_tensor` gives you uniform values `(-10, 10)` so we should have better coverage there. 

Note that this change is not to improve the gradcheck, but rather to deprecate the undocumented and untested `make_non_contiguous` from `torch.testing`.

> In particular, things like layout or memory format are not copied anymore?

- `layout` was not really functional in the first place. For example: 

    ```py
    >>> t = torch.randn(1, 2, 3).to_sparse_coo()
    >>> torch.randn_like(t)
    NotImplementedError: Could not run 'aten::normal_' with arguments from the 'SparseCPU' backend. [...]
    ```
- `memory_format` is not copied from the input tensor, but rather fixed to `torch.legacy_contiguous_format`. `make_tensor` generates contiguous tensors by default so we don't need to pass anything here.

Comment (albanD) [reply torch/autograd/gradcheck.py:1531]: > By default make_tensor gives you uniform values (-10, 10) so we should have better coverage there.

Not sure what you mean by better coverage?
There are some subtleties in the numerical precision for these functions and I would prefer the values to remain in the [-1, 1] range if possible.

> Note that this change is not to improve the gradcheck, but rather to deprecate the undocumented and untested make_non_contiguous from torch.testing.

Sure! That's not problem! I just want to make sure we're not changing the behavior of gradcheck in the process.

Comment (pmeier) [reply torch/autograd/gradcheck.py:1531]: > Not sure what you mean by better coverage?

Better coverage of different values. Happy to change `[-1, 1]` if that is required here. 

Just wondering why used a normal distribution before if we want to have this interval. Without this patch we already have about 30% of the values outside the interval.

Comment (albanD) [reply torch/autograd/gradcheck.py:1531]: I don't think there is a very strong argument for using normal vs uniform really.
The thing is that we end up doing reduction on pretty big matrices and so the increased variance can be detrimental for precision.
We do have some tests that are quite close to the numerical threshold so I don't want to start introducing flakyness.

Comment (albanD) [reply torch/autograd/gradcheck.py]: Thanks for the update!

**Review by albanD — COMMENTED** [review]
autograd update LGTM
Letting @mruberry handle the rest.

**Review by mruberry — APPROVED** [review]
Neat

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34457731).

**Pull request closed without merging**